### PR TITLE
Add Conditions Object to Statuses of AppGroup

### LINKED ISSUE #146: Display Application and ApplicationGroup reconciliation status through a list of transition states
It would be nice to capture the status as a **list of transition states** similar to how it is done in helm-controller's `HelmRelease` object status.

For example https://toolkit.fluxcd.io/components/source/helmrepositories/#status-examples
```yaml
status:
  url: http://<host>/helmrepository/default/stable/index.yaml
  conditions:
    - lastTransitionTime: "2020-04-10T09:34:45Z"
      message: Helm repository index is available at /data/helmrepository/default/stable/index-21c195d78e699e4b656e2885887d019627838993.yaml
      reason: IndexationSucceeded
      status: "True"
      type: Ready
```
> This needs some scoping

### PR DESCRIPTION
**Pull request by jonathan-innis**
Made the following changes to add conditions to the api spec so that the api spec now looks like:
```yaml
status:
  applications:
    - status:
        url: http://<host>/helmrepository/default/stable/index.yaml
        conditions: <status conditions from the underlying helmRelease object>
       - subcharts:
          - name1
             conditions:  <status conditions from the underlying helmRelease object>
          - name2
             conditions:  <status conditions from the underlying helmRelease object>
  conditions:
    - lastTransitionTime: "2020-04-10T09:34:45Z"
       message: <message>
       reason: ReconciliationSucceeded
       status: "True"
       type: Deploy
    - lastTransitionTime: "2020-04-10T09:34:45Z"
       message: <message>
       reason: WorkflowSucceeded
       status: "True"
       type: Ready
```

Closes #146

### CODE REVIEW

**Comment by nitishm** [comment]
Can you resolve the merge conflicts do that I can try this out.

**Comment by jonathan-innis** [comment]
> Can you resolve the merge conflicts do that I can try this out. 

Done

File: controllers/appgroup_controller.go [hunk 1]
```diff
@@ -165,6 +176,10 @@ func (r *ApplicationGroupReconciler) Reconcile(req ctrl.Request) (ctrl.Result, e
 	// Compares the current generation to the generation that was last
 	// seen and updated by the reconciler
 	if appGroup.Generation != appGroup.Status.ObservedGeneration {
+		// Change the app group spec into a progressing state
+		appGroup.Progressing()
+		_ = r.Status().Update(ctx, &appGroup)
```
Comment (nitishm) [inline controllers/appgroup_controller.go:181]: @jonathan-innis This is causing the `appGroup.Status.ObservedGeneration` to get bumped up.
`appGroup.Progressing()` logic below bumps it up.
```go
/ Progressing resets the conditions of the ApplicationGroup to
// metav1.Condition of type meta.ReadyCondition with status 'Unknown' and
// meta.StartingReason reason and message.
func (in *ApplicationGroup) Progressing() {
	in.Status.ObservedGeneration = in.Generation
	in.Status.Conditions = []metav1.Condition{}
	meta.SetResourceCondition(in, meta.ReadyCondition, metav1.ConditionUnknown, meta.ProgressingReason, "workflow is reconciling...")
	meta.SetResourceCondition(in, meta.DeployCondition, metav1.ConditionUnknown, meta.ProgressingReason, "application group is reconciling...")
}
```

This causes the `appGroup.Status.ObservedGeneration` to no longer be 0 on a first time reconcile which in turn enables  the `appGroup.Status.Update` flag which is checked below - leading to the following error during reconcile

```logs
{"level":"error","ts":1618417317.880827,"logger":"controllers.ApplicationGroup","msg":"failed to DELETE argo workflow object","appgroup":"bookinfo","appgroup":"bookinfo","error":"resource name may not be empty","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/nitish_malhotra/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/Azure/Orkestra/pkg/workflow.(*argo).Submit\n\t/Users/nitish_malhotra/github/azure/orkestra/pkg/workflow/argo.go:180\ngithub.com/Azure/Orkestra/controllers.(*ApplicationGroupReconciler).generateWorkflow\n\t/Users/nitish_malhotra/github/azure/orkestra/controllers/appgroup_reconciler.go:248\ngithub.com/Azure/Orkestra/controllers.(*ApplicationGroupReconciler).reconcile\n\t/Users/nitish_malhotra/github/azure/orkestra/controllers/appgroup_reconciler.go:60\ngithub.com/Azure/Orkestra/controllers.(*ApplicationGroupReconciler).Reconcile\n\t/Users/nitish_malhotra/github/azure/orkestra/controllers/appgroup_controller.go:186\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/nitish_malhotra/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/nitish_malhotra/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2\n\t/Users/nitish_malhotra/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/Users/nitish_malhotra/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/Users/nitish_malhotra/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/Users/nitish_malhotra/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/nitish_malhotra/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/Users/nitish_malhotra/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/Users/nitish_malhotra/go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99"}
```

**Review by nitishm — CHANGES_REQUESTED** [review]
Error on reconcile due to modified status.ObservedGeneration during condition update

File: controllers/appgroup_controller.go [hunk 2]
```diff
@@ -296,6 +295,8 @@ func (r *ApplicationGroupReconciler) handleResponseAndEvent(ctx context.Context,
 		errStr = err.Error()
 		grp.ReadyFailed(errStr)
 	} else {
+		// Only update the observed generation when the reconciliation succeeds
+		grp.Status.ObservedGeneration = grp.Generation
```
Comment (nitishm) [inline controllers/appgroup_controller.go:299]: This commit seems to be causing issues - things were working fine before this. Can you verify?

Comment (nitishm) [reply controllers/appgroup_controller.go:299]: NVM pilot error - bad environment. DOH!

**Review by nitishm — DISMISSED** [review]
Works perfectly
```yaml
status:
    conditions:
    - lastTransitionTime: "2021-04-15T06:11:56Z"
      message: workflow and reconciliation succeeded
      reason: Succeeded
      status: "True"
      type: Ready
    - lastTransitionTime: "2021-04-15T06:10:10Z"
      message: application group reconciliation succeeded
      reason: Succeeded
      status: "True"
      type: Deploy
    observedGeneration: 2
    status:
    - conditions:
      - lastTransitionTime: "2021-04-15T06:10:11Z"
        message: Chart fetch was successful for Helm release 'ambassador' in 'ambassador'.
        reason: ChartFetched
        status: "True"
        type: ChartFetched
      - lastTransitionTime: "2021-04-15T06:10:52Z"
        message: Installation or upgrade succeeded for Helm release 'ambassador' in
          'ambassador'.
        reason: Deployed
        status: "True"
        type: Deployed
      - lastTransitionTime: "2021-04-15T06:10:52Z"
        message: Release was successful for Helm release 'ambassador' in 'ambassador'.
        reason: Succeeded
        status: "True"
        type: Released
      name: ambassador
      phase: Succeeded
      version: 6.6.0
    - conditions:
      - lastTransitionTime: "2021-04-15T06:11:44Z"
        message: Chart fetch was successful for Helm release 'bookinfo' in 'bookinfo'.
        reason: ChartFetched
        status: "True"
        type: ChartFetched
      - lastTransitionTime: "2021-04-15T06:11:47Z"
        message: Installation or upgrade succeeded for Helm release 'bookinfo' in
          'bookinfo'.
        reason: Deployed
        status: "True"
        type: Deployed
      - lastTransitionTime: "2021-04-15T06:11:47Z"
        message: Release was successful for Helm release 'bookinfo' in 'bookinfo'.
        reason: Succeeded
        status: "True"
        type: Released
      name: bookinfo
      phase: Succeeded
      staged: true
      subcharts:
        details:
          conditions:
          - lastTransitionTime: "2021-04-15T06:11:05Z"
            message: Chart fetch was successful for Helm release 'details' in 'bookinfo'.
            reason: ChartFetched
            status: "True"
            type: ChartFetched
          - lastTransitionTime: "2021-04-15T06:11:07Z"
            message: Installation or upgrade succeeded for Helm release 'details'
              in 'bookinfo'.
            reason: Deployed
            status: "True"
            type: Deployed
          - lastTransitionTime: "2021-04-15T06:11:07Z"
            message: Release was successful for Helm release 'details' in 'bookinfo'.
            reason: Succeeded
            status: "True"
            type: Released
          staged: true
          version: v1
        productpage:
          conditions:
          - lastTransitionTime: "2021-04-15T06:11:26Z"
            message: Chart fetch was successful for Helm release 'productpage' in
              'bookinfo'.
            reason: ChartFetched
            status: "True"
            type: ChartFetched
          - lastTransitionTime: "2021-04-15T06:11:40Z"
            message: Installation or upgrade succeeded for Helm release 'productpage'
              in 'bookinfo'.
            reason: Deployed
            status: "True"
            type: Deployed
          - lastTransitionTime: "2021-04-15T06:11:41Z"
            message: Release was successful for Helm release 'productpage' in 'bookinfo'.
            reason: Succeeded
            status: "True"
            type: Released
          staged: true
          version: v1
        ratings:
          conditions:
          - lastTransitionTime: "2021-04-15T06:11:05Z"
            message: Chart fetch was successful for Helm release 'ratings' in 'bookinfo'.
            reason: ChartFetched
            status: "True"
            type: ChartFetched
          - lastTransitionTime: "2021-04-15T06:11:08Z"
            message: Installation or upgrade succeeded for Helm release 'ratings'
              in 'bookinfo'.
            reason: Deployed
            status: "True"
            type: Deployed
          - lastTransitionTime: "2021-04-15T06:11:08Z"
            message: Release was successful for Helm release 'ratings' in 'bookinfo'.
            reason: Succeeded
            status: "True"
            type: Released
          staged: true
          version: v1
        reviews:
          conditions:
          - lastTransitionTime: "2021-04-15T06:11:14Z"
            message: Chart fetch was successful for Helm release 'reviews' in 'bookinfo'.
            reason: ChartFetched
            status: "True"
            type: ChartFetched
          - lastTransitionTime: "2021-04-15T06:11:19Z"
            message: Installation or upgrade succeeded for Helm release 'reviews'
              in 'bookinfo'.
            reason: Deployed
            status: "True"
            type: Deployed
          - lastTransitionTime: "2021-04-15T06:11:19Z"
            message: Release was successful for Helm release 'reviews' in 'bookinfo'.
            reason: Succeeded
            status: "True"
            type: Released
          staged: true
          version: v1
```

**Comment by jonathan-innis** [comment]
I do need to make update to patch instead of update to prevent conflicts if someone updates spec while reconciliation is happening

**Comment by nitishm** [comment]
> I do need to make update to patch instead of update to prevent conflicts if someone updates spec while reconciliation is happening

Cool - will test it out in the AM. I pushed a commit as well to the chart with the updated CRD

File: controllers/appgroup_controller.go [hunk 3]
```diff
@@ -390,43 +381,121 @@ func (r *ApplicationGroupReconciler) handleRemediation(ctx context.Context, logr
...
-		g.Status.Phase = orkestrav1alpha1.Rollback
-		_ = r.Status().Update(ctx, &g)
-
-		return reconcile.Result{Requeue: true, RequeueAfter: requeueAfter}, nil
+		g.RollingBack()
+		_ = r.Status().Patch(ctx, &g, patch)
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
 	}
 	// Reverse and cleanup the workflow and associated helmreleases
+	g.RollingBack()
```
Comment (nitishm) [inline controllers/appgroup_controller.go:403]: @jonathan-innis Can we create a new condition like cleaning up since this isnt a rollback

**Review by nitishm — APPROVED** [review]
Create a new issue based on last comment on replacing rolling back condition with cleaning up for failures on first time install

**Pull request merged**